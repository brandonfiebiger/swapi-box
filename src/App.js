import React, { Component } from 'react';
import './App.css';

import { fetchData } from './apiCalls/apiCalls.js';
import Crawl from './Components/Crawl/Crawl.js';
import Favorites from './Components/Favorites/Favorites.js';
import Categories from './Components/Categories/Categories';
import CardContainer from './Components/CardContainer/CardContainer';

class App extends Component {
  constructor() {
    super();
    this.state = {
      categories: ['people', 'planets', 'vehicles'],
      selectedCategory: '',
      cards: [],
      favorites: [],
      films: [],
      people: [],
      planets: [],
      vehicles: [],
      loading: true
    };
  }

  async componentDidMount() {
    const cleanedData = await fetchData('films');
    this.setState({ 'films': cleanedData, loading: false });
    const storedData = localStorage.getItem("favorites");
    if (storedData) {
      var parsedData = JSON.parse(storedData);
      this.setState({ 'favorites': parsedData });
    } 
  }

  selectCategory = (category) => {
    if (!this.state[category].length) {
      this.setState({ loading: true }, async() => {
        const cleanedData = await fetchData(category);
        this.setState({ [category]: cleanedData, loading: false });
      });
    }
    this.setState({ selectedCategory: category });
  }

  selectFavorites = () => {
    if (!this.state.favorites.length) {
      alert('You currently don\'t have any favorites saved. Click a card to add or remove it from your favorites.');
    } else {
      this.setState({ selectedCategory: 'favorites' });
    }
  }

  toggleFavorites = (element) => {
    const { favorites } = this.state;
    const clickedCard = element;

    if (favorites.includes(clickedCard)) {
      const filteredFavorites = favorites.filter( favorite => (
        favorite !== clickedCard
      ));
      let dataString = JSON.stringify(filteredFavorites);
      localStorage.setItem('favorites', dataString);
      this.setState({ favorites: filteredFavorites });
    } else {
      const addedFavorites = [...favorites, clickedCard];
      let dataString = JSON.stringify(addedFavorites);
      localStorage.setItem('favorites', dataString);
      this.setState({ favorites: addedFavorites });
    }
  }

  render() {
    const { 
      favorites, 
      categories, 
      selectedCategory, 
      films, 
      loading } = this.state;
    const category = this.state[selectedCategory] || []; 
    let appLoading;
    loading ? appLoading = 'show-loading' : appLoading = 'hide-loading';
    
    return (
      <div className='App'>
        <div className={appLoading}>Loading...</div>
        <Crawl films={films} />
        <header>
          <img src={require('./images/logo.png')} alt='SWAPIbox' />
          <Categories 
            categories={categories} 
            selectedCategory={selectedCategory}
            selectCategory={this.selectCategory} />
          <Favorites 
            favorites={favorites} 
            selectFavorites={this.selectFavorites}/>
        </header>
        <main>
          <CardContainer 
            category={category} 
            favorites={favorites}
            toggleFavorites={this.toggleFavorites} />
        </main>
      </div>
    );
  }
}

export default App;