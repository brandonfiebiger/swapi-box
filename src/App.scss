@import "variables.scss";

.App {
  width: 100%;
  height: 100vh;
  font-family: $main-font;
  button {
    font-family: $main-font;
  }
}

.show-loading {
  display: flex;
  position: fixed;
  justify-content: center;
  align-items: center;
  top: 0;
  margin: 0;
  width: 100%;
  height: 100%;
  opacity: .5;
  background-color: #3a3a3a;
  color: #fff;
  font-size: 4rem;
}

.hide-loading {
  display: none;
}

header {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  width: calc(100% - 4rem);
  padding: 2rem;
}