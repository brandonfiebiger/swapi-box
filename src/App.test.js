import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {

  let wrapper;

  beforeEach(() => {
    localStorage.clear();
    wrapper = shallow(<App />);
  });

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it.skip('should check to see if there are favorites in localStorage and update state if there are', () => {
    expect(wrapper.state().favorites).toEqual([]);
    const favorites = [{name: 'test'}];

    localStorage.setItem('favorites', JSON.stringify(favorites));
    wrapper.shallow(<App />);

    expect(wrapper.state().favorites).toEqual([{name: 'test'}]);
  });
  
  describe('selectCategory', () => {

    it('should set state of selectedCategory to category being passed in', () => {
      expect(wrapper.state('selectedCategory')).toEqual('');
      wrapper.instance().selectCategory('planets');
      expect(wrapper.state('selectedCategory')).toEqual('planets');
    });
    
    it('should set the state of loading to true if the selected categories array does not have length',  () => {
      wrapper.setState({loading: false});
      wrapper.instance().selectCategory('planets');
      expect(wrapper.state('loading')).toEqual(true);
      
      wrapper.setState({loading: false});
      wrapper.setState({planets: ['moon']});
      wrapper.instance().selectCategory('planets');
      expect(wrapper.state('loading')).toEqual(false);
    });

    it('should set the state of selectedCategory to the selected category', () => {
      wrapper.instance().selectCategory('planets');
      expect(wrapper.state('selectedCategory')).toEqual('planets');
    });

  });

  describe('selectFavorites', () => {

    it('should give an alert if there are no favorites', () => {
      window.alert = jest.fn();
      wrapper.instance().selectFavorites();
      expect(window.alert).toHaveBeenCalled();
    });

    it('should update state of selectedCategory with favorites', () => {
      wrapper.setState({ favorites: [{name: 'test'}]});
      expect(wrapper.state('selectedCategory')).toEqual('');
      
      wrapper.instance().selectFavorites();
      
      expect(wrapper.state('selectedCategory')).toEqual('favorites');
    });

  });

  describe('toggleFavorites', () => {
    
    it('should update the state of favorites with a new favorite', () => {
      wrapper.instance().toggleFavorites('I love you');
      expect(wrapper.state('favorites')).toEqual(['I love you']);
    });

    it('should remove the favorite from the array when input twice', () => {
      wrapper.instance().toggleFavorites('lets be friends');
      wrapper.instance().toggleFavorites('lets be friends');
      expect(wrapper.state('favorites')).toEqual([]);
    });

    it('should save favorites to localeStorage', () => {
      const favorite = {name: 'test'};
    
      expect(localStorage.getItem('favorites')).toEqual(undefined);
  
      wrapper.instance().toggleFavorites(favorite);
  
      expect(JSON.parse(localStorage.getItem('favorites'))).toEqual([favorite]);
    });

  });

});

