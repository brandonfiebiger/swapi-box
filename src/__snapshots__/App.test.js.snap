// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `"<div class=\\"App\\"><div class=\\"show-loading\\">Loading...</div><section class=\\"crawler-container\\"><div class=\\"crawler\\"></div></section><header><img src=\\"logo.png\\"/><nav class=\\"nav\\"><button name=\\"people\\" class=\\"people btn \\">people</button><button name=\\"planets\\" class=\\"planets btn \\">planets</button><button name=\\"vehicles\\" class=\\"vehicles btn \\">vehicles</button></nav><button>View Favorites: 0</button></header><main><section></section></main></div>"`;
